rpcs3/Emu/RSX/RSXFragmentProgram.h

```diff
@@ -255,15 +255,18 @@ struct RSXFragmentProgram
 
 		data_storage_helper& operator=(const data_storage_helper& other)
 		{
-			if (other.data_ptr == other.local_storage.data())
-			{
-				local_storage = other.local_storage;
-				data_ptr = local_storage.data();
-			}
-			else
+			if (this != &other)
 			{
-				data_ptr = other.data_ptr;
-				local_storage.clear();
+				if (other.data_ptr == other.local_storage.data())
```

Comment:
These lines shouldn't have changed, do if (this == &other) return *this; instead;

Reply:
wouldn't that result in a conditional return though? could cause more problems

Reply:
You've rewritten some history for no reason. Short circuit the if. Doing so is good practice, looks better, prevents excess indenting, etc

Reply:
i dont understand what you're asking me to do ..

Reply:
i dont understand what you're asking me to do ..

Reply:
Instead of changing the whole function, just add the one line at the beginning.

Reply:
```cpp
if (this == &other)
{
	return *this;
}
if (other.data_ptr == other.local_storage.data())
...
```
Basically, if `this == &other` then just return straight away. If execution hasn't returned from that point then you can assume that this wasn't equal to &other and can put rest of the code without any additional checks.

Reply:
the else statement still modifies the other pointer, so it could still cause memory problems if you add the check to the first if

Reply:
There is no else statement. You are not understanding.

The function with your patch looks like this:
```
		data_storage_helper& operator=(const data_storage_helper& other)
		{
			if (this != &other)
			{
				if (other.data_ptr == other.local_storage.data())
				{
					local_storage = other.local_storage;
					data_ptr = local_storage.data();
				}
				else
				{
					data_ptr = other.data_ptr;
					local_storage.clear();
				}
			}

			return *this;
		}
```

Make it look like this
```
		data_storage_helper& operator=(const data_storage_helper& other)
		{
			if (this == &other) return *this;

			if (other.data_ptr == other.local_storage.data())
			{
				local_storage = other.local_storage;
				data_ptr = local_storage.data();
			}
			else
			{
				data_ptr = other.data_ptr;
				local_storage.clear();
			}

			return *this;
		}
```

Reply:
ahhh, ok :)